@xen-orchestra/fs/src/abstract.js

```diff
@@ -16,6 +15,39 @@ import { DEFAULT_ENCRYPTION_ALGORITHM, _getEncryptor } from './_encryptor'
 
 const { info, warn } = createLogger('xo:fs:abstract')
 
+function runAsync(worker, stack, concurrency = 10) {
+  return new Promise((resolve, reject) => {
+    let working = 0
+    function push(...args) {
+      stack.push(...args)
+      run()
+    }
+    async function runOne(entry) {
+      ++working
+      try {
+        await worker(entry, push)
+      } catch (error) {
+        console.warn(entry, error)
+      }
+      --working
+      run()
+    }
+    function run() {
+      while (working < concurrency) {
+        if (stack.length === 0) {
+          if (working === 0) {
+            resolve()
+          }
+          return
```

Comment:
could it lead with a premature end, if the consumption is faster than the production ?

Reply:
No because the production is done by the worker itself which means that working will not be `0`.

But this is not the definitive version, I need to add proper error handling.